File: Sources/LedgelingsCore/Hideout.swift
import Foundation

/// The little house the creatures hide in when asked to go away for a while.
///
/// away → appearing → gathering → shrinking → hidden → growing → releasing → vanishing → away
///
/// The colony drives it: it calls `entered` as each creature reaches the door,
/// acts on the events `update` returns, and draws the house at `scale`.
public struct Hideout: Equatable, Sendable {
    public enum Phase: Equatable, Sendable {
        case away, appearing, gathering, shrinking, hidden, growing, releasing, vanishing
    }

    public enum Event: Equatable, Sendable {
        /// The gathering took too long: put these creatures inside right now.
        case forceInside([Int])
        /// This creature walks out of the door.
        case letOut(Int)
    }

    public var appearTime: Double = 0.4
    public var shrinkTime: Double = 0.5
    /// Longest the house waits at the door before pulling the stragglers in.
    public var gatherCap: Double = 25
    public var releaseEvery: Double = 0.6

    public private(set) var phase: Phase = .away
    public private(set) var count = 0
    public private(set) var inside: Set<Int> = []
    private var phaseStarted: Double = 0
    private var until: Double = 0
    private var lastRelease: Double = 0

    public init() {}

    public var isActive: Bool { phase != .away }
    public func isInside(_ creature: Int) -> Bool { inside.contains(creature) }
    public func remaining(at time: Double) -> Double { max(0, until - time) }

    /// Start hiding `count` creatures for `seconds`. Ignored while already busy.
    public mutating func hide(count: Int, at time: Double, for seconds: Double) {
        guard !isActive else { return }
        self.count = count
        inside = []
        until = time + seconds
        enter(.appearing, at: time)
    }

    /// One creature reached the door.
    public mutating func entered(_ creature: Int, at time: Double) {
        guard phase == .gathering else { return }
        inside.insert(creature)
        if inside.count >= count { enter(.shrinking, at: time) }
    }

    /// "Bring them back now": open up wherever we are.
    public mutating func recall(at time: Double) {
        until = time
        switch phase {
        case .hidden: enter(.growing, at: time)
        case .shrinking: enter(.growing, at: time - appearTime * (1 - scale(at: time)))
        case .appearing, .gathering: enter(.releasing, at: time)
        default: break
        }
    }

    /// 0 = not there, 1 = full size.
    public func scale(at time: Double) -> Double {
        let sincePhase = time - phaseStarted
        switch phase {
        case .away, .hidden: return 0
        case .appearing, .growing: return min(1, max(0, sincePhase / appearTime))
        case .gathering, .releasing: return 1
        case .shrinking, .vanishing: return min(1, max(0, 1 - sincePhase / shrinkTime))
        }
    }

    public mutating func update(at time: Double) -> [Event] {
        let sincePhase = time - phaseStarted
        switch phase {
        case .away:
            return []
        case .appearing:
            if sincePhase >= appearTime { enter(.gathering, at: phaseStarted + appearTime) }
            return []
        case .gathering:
            if inside.count >= count { enter(.shrinking, at: time); return [] }
            guard sincePhase >= gatherCap else { return [] }
            return [.forceInside((0..<count).filter { !inside.contains($0) })]
        case .shrinking:
            if sincePhase >= shrinkTime { enter(.hidden, at: time) }
            return []
        case .hidden:
            if time >= until { enter(.growing, at: time) }
            return []
        case .growing:
            if sincePhase >= appearTime { enter(.releasing, at: time) }
            return []
        case .releasing:
            guard let next = inside.min() else { enter(.vanishing, at: time); return [] }
            guard time - lastRelease >= releaseEvery else { return [] }
            inside.remove(next)
            lastRelease = time
            return [.letOut(next)]
        case .vanishing:
            if sincePhase >= shrinkTime { enter(.away, at: time) }
            return []
        }
    }

    private mutating func enter(_ next: Phase, at time: Double) {
        phase = next
        phaseStarted = time
        if next == .releasing { lastRelease = time - releaseEvery }     // the first one steps out at once
    }
}

File: win/LedgelingsCore/Hideout.cs
namespace Ledgelings.Core;

/// <summary>
/// The little house the creatures hide in when asked to go away for a while.
///
/// away → appearing → gathering → shrinking → hidden → growing → releasing → vanishing → away
///
/// The colony drives it: it calls <see cref="Entered"/> as each creature reaches the door,
/// acts on the events <see cref="Update"/> returns, and draws the house at <see cref="Scale"/>.
/// </summary>
public sealed class Hideout
{
    public enum Phase { Away, Appearing, Gathering, Shrinking, Hidden, Growing, Releasing, Vanishing }

    public abstract record Event
    {
        /// <summary>The gathering took too long: put these creatures inside right now.</summary>
        public sealed record ForceInside(IReadOnlyList<int> Creatures) : Event
        {
            public bool Equals(ForceInside? other) => other is not null && Creatures.SequenceEqual(other.Creatures);
            public override int GetHashCode() => Creatures.Count;
        }
        /// <summary>This creature walks out of the door.</summary>
        public sealed record LetOut(int Creature) : Event;
    }

    public double AppearTime { get; set; } = 0.4;
    public double ShrinkTime { get; set; } = 0.5;
    /// <summary>Longest the house waits at the door before pulling the stragglers in.</summary>
    public double GatherCap { get; set; } = 25;
    public double ReleaseEvery { get; set; } = 0.6;

    public Phase CurrentPhase { get; private set; } = Phase.Away;
    public int Count { get; private set; }
    private readonly SortedSet<int> inside = new();
    public IReadOnlySet<int> Inside => inside;
    private double phaseStarted;
    private double until;
    private double lastRelease;

    public bool IsActive => CurrentPhase != Phase.Away;
    public bool IsInside(int creature) => inside.Contains(creature);
    public double Remaining(double time) => Math.Max(0, until - time);

    /// <summary>Start hiding <paramref name="count"/> creatures for <paramref name="seconds"/>. Ignored while already busy.</summary>
    public void Hide(int count, double time, double seconds)
    {
        if (IsActive) return;
        Count = count;
        inside.Clear();
        until = time + seconds;
        Enter(Phase.Appearing, time);
    }

    /// <summary>One creature reached the door.</summary>
    public void Entered(int creature, double time)
    {
        if (CurrentPhase != Phase.Gathering) return;
        inside.Add(creature);
        if (inside.Count >= Count) Enter(Phase.Shrinking, time);
    }

    /// <summary>"Bring them back now": open up wherever we are.</summary>
    public void Recall(double time)
    {
        until = time;
        switch (CurrentPhase)
        {
            case Phase.Hidden: Enter(Phase.Growing, time); break;
            case Phase.Shrinking: Enter(Phase.Growing, time - AppearTime * (1 - Scale(time))); break;
            case Phase.Appearing:
            case Phase.Gathering: Enter(Phase.Releasing, time); break;
        }
    }

    /// <summary>0 = not there, 1 = full size.</summary>
    public double Scale(double time)
    {
        var sincePhase = time - phaseStarted;
        return CurrentPhase switch
        {
            Phase.Away or Phase.Hidden => 0,
            Phase.Appearing or Phase.Growing => Math.Min(1, Math.Max(0, sincePhase / AppearTime)),
            Phase.Gathering or Phase.Releasing => 1,
            _ => Math.Min(1, Math.Max(0, 1 - sincePhase / ShrinkTime)),
        };
    }

    public List<Event> Update(double time)
    {
        var events = new List<Event>();
        var sincePhase = time - phaseStarted;
        switch (CurrentPhase)
        {
            case Phase.Away:
                break;
            case Phase.Appearing:
                if (sincePhase >= AppearTime) Enter(Phase.Gathering, phaseStarted + AppearTime);
                break;
            case Phase.Gathering:
                if (inside.Count >= Count) { Enter(Phase.Shrinking, time); break; }
                if (sincePhase < GatherCap) break;
                events.Add(new Event.ForceInside(Enumerable.Range(0, Count).Where(i => !inside.Contains(i)).ToList()));
                break;
            case Phase.Shrinking:
                if (sincePhase >= ShrinkTime) Enter(Phase.Hidden, time);
                break;
            case Phase.Hidden:
                if (time >= until) Enter(Phase.Growing, time);
                break;
            case Phase.Growing:
                if (sincePhase >= AppearTime) Enter(Phase.Releasing, time);
                break;
            case Phase.Releasing:
                if (inside.Count == 0) { Enter(Phase.Vanishing, time); break; }
                if (time - lastRelease < ReleaseEvery) break;
                var next = inside.Min;
                inside.Remove(next);
                lastRelease = time;
                events.Add(new Event.LetOut(next));
                break;
            case Phase.Vanishing:
                if (sincePhase >= ShrinkTime) Enter(Phase.Away, time);
                break;
        }
        return events;
    }

    private void Enter(Phase next, double time)
    {
        CurrentPhase = next;
        phaseStarted = time;
        if (next == Phase.Releasing) lastRelease = time - ReleaseEvery;     // the first one steps out at once
    }
}
